File: 2025/4.py
# python 4.py < input.txt  0.24s user 0.02s system 90% cpu 0.284 total

import sys

def count_isolated_at_symbols(lines, H, W):
    ans = 0
    at_positions = []

    for r in range(H):
        for c in range(W):
            if lines[r][c] == '@':
                # Found an '@', count adjacent '@' symbols
                adjacent_at_counts = 0

                # Check all 8 adjacent cells
                for dr in range(-1, 2):
                    for dc in range(-1, 2):
                        if dr == 0 and dc == 0:
                            continue
                        nr = r + dr
                        nc = c + dc

                        if 0 <= nr < H and 0 <= nc < W:
                            if lines[nr][nc] == '@':
                                adjacent_at_counts += 1

                # If fewer than 4 adjacent '@' symbols, this is "isolated"
                if adjacent_at_counts < 4:
                    at_positions.append((r, c))
                    ans += 1

    # Remove the '@' symbols from the grid
    for r, c in at_positions:
        lines[r][c] = '.'

    return ans

def main():
    lines = []
    for line in sys.stdin:
        line = line.rstrip('\n')
        if line:
            lines.append(list(line))  # Convert to list for mutability

    if not lines:
        return

    H = len(lines)
    W = len(lines[0])

    ans = count_isolated_at_symbols(lines, H, W)

    print(ans)

    # Keep processing until no more isolated '@' symbols
    while True:
        new_ans = count_isolated_at_symbols(lines, H, W)
        if new_ans == 0:
            break
        ans += new_ans

    print(ans)

if __name__ == "__main__":
    main()

File: 2025/4.cpp
// ./a.out < input.txt  0.04s user 0.00s system 85% cpu 0.046 total

#include <iostream>
#include <string>
#include <vector>
using namespace std;

long long countIsolatedAtSymbols(const vector<string> &lines, int H, int W)
{
    long long ans = 0;
    vector<pair<int, int>> atPositions;

    for (int r = 0; r < H; ++r)
    {
        for (int c = 0; c < W; ++c)
        {
            if (lines[r][c] == '@')
            {
                // Found an '@', look for adjacent numbers
                long long adjacentAtCounts = 0;

                // Count adjacent @ to the current '@'
                // if fewer than 3, count this '@', otherwise skip them.

                for (int dr = -1; dr <= 1; ++dr)
                {
                    for (int dc = -1; dc <= 1; ++dc)
                    {
                        if (dr == 0 && dc == 0)
                            continue;
                        int nr = r + dr;
                        int nc = c + dc;

                        if (nr >= 0 && nr < H && nc >= 0 && nc < W)
                        {
                            if (lines[nr][nc] == '@')
                            {
                                adjacentAtCounts++;
                            }
                        }
                    }
                }

                if (adjacentAtCounts < 4)
                {
                    atPositions.push_back({r, c});
                    ans++;
                }
            }
        }
    }

    for(const auto &pos : atPositions) {
        int r = pos.first;
        int c = pos.second;
        // Remove the '@' from the grid
        // to avoid counting it again in subsequent calls
        const_cast<string&>(lines[r])[c] = '.'; 
    }
    return ans;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<string> lines;
    string line;
    while (getline(cin, line))
    {
        if (line.empty())
            continue;
        lines.push_back(line);
    }
    if (lines.empty())
        return 0;

    int H = lines.size();
    int W = lines[0].size();

    long long ans = countIsolatedAtSymbols(lines, H, W);

    cout << ans << "\n";

    while(true) {
        long long newAns = countIsolatedAtSymbols(lines, H, W);
        if(newAns == 0) {
            break;
        }
        ans += newAns;
    }

    cout << ans << "\n";
    return 0;
}